README.md
# AnyQuestion

[![Build Status](https://img.shields.io/travis/mose/anyquestion.svg)](https://travis-ci.org/mose/anyquestion)


This is a web app for managing the step "Do you have any question?" that happens at the end of a talk. It's inspired by a feature present in [Daskit](https://www.daskit.com/) which is not open source (yet?). Every talk has a room, people ask questions there and vote on other people questions. Then it gets easy for the speaker to select the question with the more weight.

There is a test install on http://anyquestion.herokuapp.com/ . As the data are not persisted, it's reset when the dyno goes to sleep.

**Warning**: this piece of code is designed for a specific use case that doesn't require high security but just a tool for quick (and dirty) polling on questions after talks. It has not been tested at scale, and its security setup is quite minimal.

## Usage

You have been to conferences, right? Every time at then end, there is some time for the questions. But sometimes people are shy, or there is some mike issue, or people are not native speaker and their strong accent prevent everybody to understand what they ask.

Make this web app reachable for you attendees. They can add questions during the talk, so there is no way to forget them. They can upvote the questions of other people. They also can cancel their own upvotes if they change their minds or if the speaker answered the question before it was asked, in the progress of his talk. Questions that reach zero vote will just disappear.

Typically, at the end, the speaker or someone else can access the questions, pick in the ones that are the most popular and read them. The speaker can then answer them.

## Dev

Code is not finished yet. but you can launch with:

    shards install

    # dev
    crystal src/anyquestion.cr

    # prod
    crystal build --release src/anyquestion.cr
    ./anyquestion -e production

    # deploy on heroku
    heroku create appname --buildpack https://github.com/crystal-lang/heroku-buildpack-crystal.git
    heroku git:remote -a appname
    git push heroku master

## dev links

- http://crystal-lang.org/docs/installation/index.html
- http://crystal-lang.org/docs
- http://crystal-lang.org/api
- http://kemalcr.com/

## Todo

Check the [CHANGELOG.md](CHANGELOG.md) file for details.

- <s>debug react/kemal for it to refresh always, for now sometimes it doesn't refresh after new question</s>
- <s>add a possibility to cancel an upvote</s>
- <s>add a config file/env vars system</s>
- <s>add an admin account for editing things and cleaning up (oauth2 to github probably)</s>
- <s>consider an admin action to dump existing set of questions in case of restart (export/import)</s>
- add some damn tests
- expose some api endpoints for plugging irc/slack bots into this
- use fibers (`spawn`) to handle websockets maybe
- make a debian package (and redhat maybe, see https://github.com/waghanza/plunder)
- check responsivity


## Contributing

- fork it, twist it, submit a PR.

## License

Copyright (c) 2016 mose - MIT license

## Contributors

- [mose](https://github.com/mose) - author


CHANGELOG.md
AnyQuestion Changelog
=========================

### v0.1.6 - wip
- link with travis
- write more tests
- upgrade kemal to 0.14.0

### v0.1.5 - 2016-06-26
- add autologged config param for easing up development
- add html escape for questions
- rename 'room' by 'talk'
- begin to add some tests

### v0.1.4 - 2016-06-25
- import/export saves and restores all the rooms

### v0.1.3 - 2016-06-25
- make possible to override config values with environment variables
- add clean features for rooms and questions

### v0.1.2 - 2016-06-23
- add an admin login option
- add a way to restrict room creation to admin
- add more config parameters for doing so

### v0.1.1 - 2016-06-17
- add possibility to downvote a previously upvote question
- questions reaching zero disappear (effectively enabling cancelling of questions)

### v0.1.0 - 2016-06-16
- fix refresh, making the app totally usable on real cases

### v0.0.2 - 2016-06-15
- initial working version

### v0.0.1 - 2016-06-08
- first line of code


public/js/room.js
'use strict';

Array.prototype.contains = function(obj) {
  var i = this.length;
  while (i--) {
    if (this[i] === obj) {
      return true;
    }
  }
  return false;
}

var User = function(){
  var user = parseInt(localStorage.getItem('aq_userid')) || Math.floor((Math.random() * 100000000) + 1);
  localStorage.setItem('aq_userid', user);
  return user;
}

var Server = function(){
  var ws_protocol = location.protocol == "https:" ? 'wss' : 'ws';
  return new WebSocket(`${ws_protocol}://${location.hostname}:${location.port}/ws?room=${location.pathname.split('/').reverse()[0]}`);
}

var Room = React.createClass({

  getInitialState: function () {
    return {
      question: '',
      questions: []
    }
  },

  componentDidMount: function () {
    this.sendable = true;
    var self = this;
    var server, user, questions;

    server = Server();
    user = User();

    server.onmessage = event => {
      questions = JSON.parse(event.data);
      self.setState({questions});
      self.refs.question.focus();
    };

    server.onopen = () => {
      server.send("hi");
    };

    server.onclose = (e) => {
      console.log(e);
      server.send("bye");
    };

    this.props.server = server;
    this.props.user = user;

    this.refs.question.focus();
  },

  componentWillUnmount: function () {
    window.removeEventListener("unload");
  },

  sendQuestion: function () {
    console.log(this.props.server.readyState);
    var self = this;
    if (!this.sendable || this.refs.question.value === "") {
      return false;
    }

    setTimeout(() => {
      self.sendable = true;
    }, 100);

    this.props.server.send(`${this.props.user}::::${this.refs.question.value}`);

    this.refs.question.value = '';
    this.sendable = false;
  },

  sendVote: function (qid) {
    this.props.server.send(`${qid}----${this.props.user}`);
  },

  sendQuestionWithEnter: function (e) {
    if (e.keyCode == 13) {
     this.sendQuestion();
    }
  },

  render: function () {
    var user = this.props.user;
    var server = this.props.server;

    const questions = this.state.questions.map(q => React.createElement("li", null,
      React.createElement('span', { className: (q.voters.contains(user) ? "voted" : "votable"), onClick: this.sendVote.bind(this, q.id) }, q.voters.length),
      React.createElement('span', { className: "q" }, q.name)
    ));

    return React.createElement("div", { className: "questions" },
      React.createElement("input", { autofocus: true, placeholder: "What is your question?", type: "text", ref: "question", onKeyUp: this.sendQuestionWithEnter.bind(this) }),
      React.createElement("ul", null, questions)
    );
  }
});

ReactDOM.render(React.createElement(Room, null), document.getElementById('room'));


public/js/talk.js
'use strict';

Array.prototype.contains = function(obj) {
  var i = this.length;
  while (i--) {
    if (this[i] === obj) {
      return true;
    }
  }
  return false;
}

var User = function(){
  var user = parseInt(localStorage.getItem('aq_userid')) || Math.floor((Math.random() * 100000000) + 1);
  localStorage.setItem('aq_userid', user);
  return user;
}

var Server = function(){
  var ws_protocol = location.protocol == "https:" ? 'wss' : 'ws';
  return new WebSocket(`${ws_protocol}://${location.hostname}:${location.port}/ws?talk=${location.pathname.split('/').reverse()[0]}`);
}

var Talk = React.createClass({

  getInitialState: function () {
    return {
      question: '',
      questions: []
    }
  },

  componentDidMount: function () {
    this.sendable = true;
    var self = this;
    var server, user, questions;

    server = Server();
    user = User();

    server.onmessage = event => {
      questions = JSON.parse(event.data);
      self.setState({questions});
      self.refs.question.focus();
    };

    server.onopen = () => {
      server.send("hi");
    };

    server.onclose = (e) => {
      console.log(e);
      server.send("bye");
    };

    this.props.server = server;
    this.props.user = user;

    this.refs.question.focus();
  },

  componentWillUnmount: function () {
    window.removeEventListener("unload");
  },

  sendQuestion: function () {
    console.log(this.props.server.readyState);
    var self = this;
    if (!this.sendable || this.refs.question.value === "") {
      return false;
    }

    setTimeout(() => {
      self.sendable = true;
    }, 100);

    this.props.server.send(`${this.props.user}::::${this.refs.question.value}`);

    this.refs.question.value = '';
    this.sendable = false;
  },

  sendVote: function (qid) {
    this.props.server.send(`${qid}----${this.props.user}`);
  },

  sendQuestionWithEnter: function (e) {
    if (e.keyCode == 13) {
     this.sendQuestion();
    }
  },

  render: function () {
    var user = this.props.user;
    var server = this.props.server;

    const questions = this.state.questions.map(q => React.createElement("li", null,
      React.createElement('span', { className: (q.voters.contains(user) ? "voted" : "votable"), onClick: this.sendVote.bind(this, q.id) }, q.voters.length),
      React.createElement('span', { className: "q" }, q.name)
    ));

    return React.createElement("div", { className: "questions" },
      React.createElement("input", { autofocus: true, placeholder: "What is your question?", type: "text", ref: "question", onKeyUp: this.sendQuestionWithEnter.bind(this) }),
      React.createElement("ul", null, questions)
    );
  }
});

ReactDOM.render(React.createElement(Talk, null), document.getElementById('talk'));
